### PR TITLE
chore(ppai/timeseries-classification): add global fishing watch sample

### PR DESCRIPTION
**Pull request by davidcavazos**
## Description

Add Global Fishing Watch sample for the People & Planet AI Series.

Staged:

* [Notebook](https://colab.research.google.com/github/davidcavazos/python-docs-samples/blob/ppai-timeseries-classification/people-and-planet-ai/timeseries-classification/README.ipynb) -- this serves as the README
  > ⚠️ When cloning the repo, it doesn't exist yet in the main branch, so replace with:
  > ```
  > !git clone --branch ppai-timeseries-classification https://github.com/davidcavazos/python-docs-samples.git ~/python-docs-samples
  > ```
  > The links to source files in the notebook will also be broken since the files don't exist yet in the main branch.

## Checklist
- [x] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/AUTHORING_GUIDE.md)
- [x] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/AUTHORING_GUIDE.md#readme-file)
- [x] **Tests** pass:   `nox -s py-3.6` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/AUTHORING_GUIDE.md#test-environment-setup))
- [x] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/AUTHORING_GUIDE.md#test-environment-setup))
- [x] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [x] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] Please **merge** this PR for me once it is approved.
- [x] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/.github/CODEOWNERS) with the codeowners for this sample

### CODE REVIEW

**Comment by dinagraves** [comment]
@davidcavazos Let me know if you need help getting this to pass!

**Comment by davidcavazos** [comment]
Thanks! It looks like some of the gcloud commands need a little tweaking, maybe some defaults changed with the latest gcloud versions? Hopefully it shouldn't take long for the tests to pass.

The code itself should be ready for review, I'm only fixing some minor things in the tests. The README will be replaced with a notebook, but that's still in process.

**Comment by davidcavazos** [comment]
Hi @nikitamaia and @loferris, can you help me review this? Thanks!

File: people-and-planet-ai/timeseries-classification/trainer.py
Comment (nikitamaia) [inline people-and-planet-ai/timeseries-classification/trainer.py:219]: My recommendation would be to pass in the number of epochs instead of the steps per epoch. Keras will calculate the correct number of steps (num samples/batch size). You don't need to repeat the dataset in this case either. When you factor in distributed training and start scaling the batch size, your steps per epoch will change as well since you're now seeing more data in each step. IMO passing the epoch and letting keras do the calculation is just less complexity.

**Comment by loferris** [comment]
I'm seeing if I can upload and run the `README.ipynb` notebook on my own; it might be nice if there were a link to open the notebook in Colab directly from the file. Grabbing the Colab button from our in progress template might work (https://github.com/GoogleCloudPlatform/bigquery-notebooks/blob/notebook_template/notebooks/official/notebook_template.ipynb). current sources of inspiration are both TensorFlow and Vertex AI docs.

**Comment by davidcavazos** [comment]
@loferris Sure, I can add a button to open in Colab. Note that the link won't work until the notebook is merged since it doesn't exist yet in the main branch. I included a link to the "staged" notebook in the PR description at the very top:

https://colab.research.google.com/github/davidcavazos/python-docs-samples/blob/ppai-timeseries-classification/people-and-planet-ai/timeseries-classification/README.ipynb

File: people-and-planet-ai/timeseries-classification/trainer.py
Comment (davidcavazos) [reply people-and-planet-ai/timeseries-classification/trainer.py]: Thank you! This is awesome. I had some trouble, but in the end I still needed to pass `eval_steps` for validation, even if the training used epochs. If I didn't pass `eval_steps` I got an error of having "empty" data points. If you know of a better way to do it, please let me know! But having it this way, at least tests are passing again.

**Comment by davidcavazos** [comment]
@nikitamaia, @loferris, tests are passing, I think it might be ready to merge unless you have any other comments.

**Comment by loferris** [comment]
> @nikitamaia, @loferris, tests are passing, I think it might be ready to merge unless you have any other comments.

LGTM!

**Comment by nikitamaia** [comment]
LGTM

**Comment by davidcavazos** [comment]
Hi @busunkim96 and @dinagraves, all tests are passing now and we've got the LGTMs from other team members in the project. Please let me know if you have any comments on it, otherwise I think it should be ready to merge :)

File: people-and-planet-ai/timeseries-classification/data_utils.py
Comment (dinagraves) [inline people-and-planet-ai/timeseries-classification/data_utils.py:50]: In the e2e test it says:

`UserWarning: Input dict contained keys ['timestamp', 'mmsi', 'distance_from_shore'] which did not match any model input. They will be ignored by the model.`

Is this expected?

Comment (davidcavazos) [reply people-and-planet-ai/timeseries-classification/data_utils.py:50]: It's a harmless warning that we were passing extra inputs to the model. But good catch on it, I made a slight modification to only pass the inputs the model is expecting to get rid of this warning.

**Comment by davidcavazos** [comment]
I just updated the nox config file to ignore Python 3.10, but the build still looks like it's failing. Otherwise, all tests are passing.

**Comment by davidcavazos** [comment]
It looks like dropping the last batch with possibly incomplete number of elements solved the flaky issue.

**Comment by davidcavazos** [comment]
@dinagraves @busunkim96 -- tests are passing now, they should no longer be flaky.

File: people-and-planet-ai/timeseries-classification/e2e_test.py
Comment (busunkim96) [inline people-and-planet-ai/timeseries-classification/e2e_test.py:267]: Since you're making multiple requests with the same header I recommend `google.auth.transport.requests.AuthorizedSession`. This will refresh the credential pre-request as needed and attach the auth header.

I believe it will also let you remove the `access_token` fixture that calls out to gcloud.

https://googleapis.dev/python/google-auth/latest/reference/google.auth.transport.requests.html#google.auth.transport.requests.AuthorizedSession

```py
import google.auth
from google.auth.transport.requests import AuthorizedSession

credentials, _ = google.auth.default()
authed_session = AuthorizedSession(credentials)
response = authed_session.post(url="...", json="...")
```

File: people-and-planet-ai/timeseries-classification/main.py
Comment (busunkim96) [inline people-and-planet-ai/timeseries-classification/main.py:90]: Can this exception be made more specific? Is it possible for the try/except block to be smaller?

https://google.github.io/styleguide/pyguide.html#244-decision

**Review by busunkim96 — APPROVED** [review]

File: people-and-planet-ai/timeseries-classification/main.py
Comment (davidcavazos) [reply people-and-planet-ai/timeseries-classification/main.py:90]: Great point, I'm also not a fan of open exceptions like this, but in this case I think it falls under one of the exceptions:

```
Never use catch-all except: statements, or catch Exception or StandardError, unless you are

* re-raising the exception, or
* creating an isolation point in the program where exceptions are not propagated but are
  recorded and suppressed instead, such as protecting a thread from crashing by guarding
  its outermost block.
```

In this case, all the exception catching in `main.py` is surrounding the entire underlying call to avoid having the server instance crash, and so we actually get a more sensible error message in the client side other than a 500 error. Ideally, each underlying call should handle its more specific exceptions.

File: people-and-planet-ai/timeseries-classification/e2e_test.py
Comment (davidcavazos) [reply people-and-planet-ai/timeseries-classification/e2e_test.py:267]: Nice! I didn't knew this existed! I'll give it a try! :)

Comment (davidcavazos) [reply people-and-planet-ai/timeseries-classification/e2e_test.py:267]: I got an error:

`'WWW-Authenticate': 'Bearer error="invalid_token" error_description="The access token could not be verified"'`

Cloud Run might be looking for something different in the request.

Comment (davidcavazos) [reply people-and-planet-ai/timeseries-classification/e2e_test.py:267]: I think the authenticated session might be using an "access token" while the Cloud Run service requires an "identity token". The notebook links to the
[Authentication overview](https://cloud.google.com/run/docs/authenticating/overview)
page for Cloud Run, which explains different methods of authenticating.

https://stackoverflow.com/questions/68655602/gcloud-identity-token-vs-access-token

**Comment by busunkim96** [comment]
Admin Merging - 

@davidcavazos is a CODEOWNER for this product area and the PR was also reviewed by @nikitamaia.

The 3.10 failure is expected as that job is not yet fully configured, see #6762.

The 3.8 failure is due to a timeout that occurred after the tests passed. (see [log](https://source.cloud.google.com/results/invocations/35fb8920-b8e5-471f-b58b-cbfa3cda85f4))

```
============================= test session starts ==============================
platform linux -- Python 3.8.8, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /workspace/people-and-planet-ai/timeseries-classification/.nox/py-3-8/bin/python
cachedir: .pytest_cache
rootdir: /workspace, configfile: pytest.ini
plugins: forked-1.3.0, xdist-2.3.0
gw0 I / gw1 I / gw2 I / gw3 I / gw4 I / gw5 I / gw6 I / gw7 I

[gw0] linux Python 3.8.8 cwd: /workspace/people-and-planet-ai/timeseries-classification

[gw1] linux Python 3.8.8 cwd: /workspace/people-and-planet-ai/timeseries-classification

[gw2] linux Python 3.8.8 cwd: /workspace/people-and-planet-ai/timeseries-classification

[gw3] linux Python 3.8.8 cwd: /workspace/people-and-planet-ai/timeseries-classification

[gw4] linux Python 3.8.8 cwd: /workspace/people-and-planet-ai/timeseries-classification

[gw5] linux Python 3.8.8 cwd: /workspace/people-and-planet-ai/timeseries-classification

[gw6] linux Python 3.8.8 cwd: /workspace/people-and-planet-ai/timeseries-classification

[gw7] linux Python 3.8.8 cwd: /workspace/people-and-planet-ai/timeseries-classification

[gw0] Python 3.8.8 (default, Mar 26 2021, 10:10:50)  -- [GCC 9.3.0]

[gw1] Python 3.8.8 (default, Mar 26 2021, 10:10:50)  -- [GCC 9.3.0]

[gw2] Python 3.8.8 (default, Mar 26 2021, 10:10:50)  -- [GCC 9.3.0]

[gw3] Python 3.8.8 (default, Mar 26 2021, 10:10:50)  -- [GCC 9.3.0]

[gw4] Python 3.8.8 (default, Mar 26 2021, 10:10:50)  -- [GCC 9.3.0]

[gw5] Python 3.8.8 (default, Mar 26 2021, 10:10:50)  -- [GCC 9.3.0]

[gw6] Python 3.8.8 (default, Mar 26 2021, 10:10:50)  -- [GCC 9.3.0]

[gw7] Python 3.8.8 (default, Mar 26 2021, 10:10:50)  -- [GCC 9.3.0]
gw0 [7] / gw1 [7] / gw2 [7] / gw3 [7] / gw4 [7] / gw5 [7] / gw6 [7] / gw7 [7]

scheduling tests via LoadScheduling

local_test.py::test_validated_incompatible_shape 
local_test.py::test_validated_ok 
e2e_test.py::test_predict 
local_test.py::test_validated_incompatible_type 
[gw3] [ 14%] PASSED local_test.py::test_validated_incompatible_shape 
local_test.py::test_serialize_deserialize 
[gw2] [ 28%] PASSED local_test.py::test_validated_incompatible_type 
local_test.py::test_e2e_local 
[gw4] [ 42%] PASSED local_test.py::test_validated_ok 
local_test.py::test_validated_missing_field 
[gw1] [ 57%] PASSED local_test.py::test_validated_missing_field 
[gw5] [ 71%] PASSED local_test.py::test_serialize_deserialize 
[gw6] [ 85%] PASSED local_test.py::test_e2e_local 
[gw0] [100%] PASSED e2e_test.py::test_predict 

=============================== warnings summary ===============================
.nox/py-3-8/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py:22
.nox/py-3-8/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py:22
.nox/py-3-8/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py:22
.nox/py-3-8/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py:22
.nox/py-3-8/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py:22
.nox/py-3-8/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py:22
.nox/py-3-8/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py:22
.nox/py-3-8/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py:22
  /workspace/people-and-planet-ai/timeseries-classification/.nox/py-3-8/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py:22: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:696
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:696
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:696
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:696
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:696
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:696
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:696
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:696
  /workspace/people-and-planet-ai/timeseries-classification/.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:696: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working
    if not isinstance(type_params, collections.Iterable):

.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:535
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:535
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:535
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:535
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:535
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:535
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:535
.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:535
  /workspace/people-and-planet-ai/timeseries-classification/.nox/py-3-8/lib/python3.8/site-packages/apache_beam/typehints/typehints.py:535: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working
    if not isinstance(type_params, (collections.Sequence, set)):

-- Docs: https://docs.pytest.org/en/stable/warnings.html
- generated xml file: /workspace/people-and-planet-ai/timeseries-classification/sponge_log.xml -
================= 7 passed, 24 warnings in 1733.28s (0:28:53) ==================
nox > Session py-3.8 was successful.


err: signal: killed


#
# Summary 
#

SUCCESS: 0, FAILURE: 1, SKIPPED: 0, ERROR: 0, 
people-and-planet-ai/timeseries-classification........................[ FAILURE]
================================================================
2021-09-29T19:54:27Z: Build finished with 2 
================================================================
```

**Pull request merged**
